albatross/tune.h

```diff
@@ -20,76 +20,111 @@
 
 namespace albatross {
 
-std::vector<ParameterValue>
+inline std::vector<ParameterValue>
 transform_parameters(const std::vector<ParameterValue> &x) {
   std::vector<ParameterValue> transformed(x.size());
   std::transform(x.begin(), x.end(), transformed.begin(), log);
   return transformed;
 }
 
-std::vector<ParameterValue>
+inline std::vector<ParameterValue>
 inverse_parameters(const std::vector<ParameterValue> &x) {
   std::vector<ParameterValue> inverted(x.size());
   std::transform(x.begin(), x.end(), inverted.begin(), exp);
   return inverted;
 }
 
-template <class Predictor>
-using TuningMetric = std::function<double(const RegressionDataset<Predictor> &,
-                                          RegressionModel<Predictor> *)>;
+template <class FeatureType>
+using TuningMetric = std::function<double(
+    const RegressionDataset<FeatureType> &, RegressionModel<FeatureType> *)>;
 
-template <class Predictor> struct TuneModelConfg {
-  RegressionModelCreator<Predictor> model_creator;
-  RegressionDataset<Predictor> dataset;
-  TuningMetric<Predictor> metric;
+using TuningMetricAggregator =
+    std::function<double(const std::vector<double> &metrics)>;
 
-  TuneModelConfg(const RegressionModelCreator<Predictor> &model_creator_,
-                 const RegressionDataset<Predictor> &dataset_,
-                 const TuningMetric<Predictor> &metric_)
-      : model_creator(model_creator_), dataset(dataset_), metric(metric_){};
+/*
+ * Returns the mean of metrics computed across multiple datasets.
+ */
+inline double mean_aggregator(const std::vector<double> &metrics) {
+  double mean = 0.;
+  for (const auto &metric : metrics) {
+    mean += metric;
+  }
+  mean /= static_cast<double>(metrics.size());
+  return mean;
+}
+
+template <class FeatureType> struct TuneModelConfg {
+  RegressionModelCreator<FeatureType> model_creator;
+  std::vector<RegressionDataset<FeatureType>> datasets;
+  TuningMetric<FeatureType> metric;
+  TuningMetricAggregator aggregator;
+  std::ostream &output_stream;
+
+  TuneModelConfg(const RegressionModelCreator<FeatureType> &model_creator_,
+                 const RegressionDataset<FeatureType> &dataset_,
+                 const TuningMetric<FeatureType> &metric_,
+                 const TuningMetricAggregator &aggregator_ = mean_aggregator,
+                 std::ostream &output_stream_ = std::cout)
+      : model_creator(model_creator_), datasets({dataset_}), metric(metric_),
+        aggregator(aggregator_), output_stream(output_stream_){};
+
+  TuneModelConfg(const RegressionModelCreator<FeatureType> &model_creator_,
+                 const std::vector<RegressionDataset<FeatureType>> &datasets_,
+                 const TuningMetric<FeatureType> &metric_,
+                 const TuningMetricAggregator &aggregator_ = mean_aggregator,
```

Comment:
Do we want to use a `mean_aggregator` over all the metrics? Seems like using median might be more robust here.

Reply:
Yeah I think median would be more robust, but that is probably not desirable for a default tuning metric.  For example, one might be tuning using negative log likelihood and with the median you may end up with some model that performs better on 51% of the data but really horribly on 49%.  I think in general such a model would be less desirable than one that performed well on average.